internal/gatewayapi/route.go

```diff
@@ -980,28 +982,83 @@ func (t *Translator) processDestEndpoints(backendRef v1beta1.BackendRef,
 		return nil, weight
 	}
 
-	var backendIps []string
 	switch KindDerefOr(backendRef.Kind, KindService) {
+	// TODO: Use EndpointSlice for ServiceImport
 	case KindServiceImport:
-		backendIps = resources.GetServiceImport(backendNamespace, string(backendRef.Name)).Spec.IPs
+		backendIps := resources.GetServiceImport(backendNamespace, string(backendRef.Name)).Spec.IPs
+		for _, ip := range backendIps {
+			var ep *ir.DestinationEndpoint
+			// Weights are not relevant for TCP and UDP Routes
+			if routeType == KindTCPRoute || routeType == KindUDPRoute {
+				ep = ir.NewDestEndpoint(
+					ip,
+					uint32(*backendRef.Port))
+			} else {
+				ep = ir.NewDestEndpointWithWeight(
+					ip,
+					uint32(*backendRef.Port),
+					weight)
+			}
+			endpoints = append(endpoints, ep)
+		}
 	case KindService:
-		backendIps = []string{resources.GetService(backendNamespace, string(backendRef.Name)).Spec.ClusterIP}
-	}
+		service := resources.GetService(backendNamespace, string(backendRef.Name))
+		var servicePort corev1.ServicePort
+		for _, port := range service.Spec.Ports {
+			if port.Port == int32(*backendRef.Port) {
+				servicePort = port
+				break
+			}
+		}
 
-	for _, ip := range backendIps {
-		var ep *ir.DestinationEndpoint
-		// Weights are not relevant for TCP and UDP Routes
-		if routeType == KindTCPRoute || routeType == KindUDPRoute {
-			ep = ir.NewDestEndpoint(
-				ip,
-				uint32(*backendRef.Port))
+		// Route to endpoints by default
+		if !t.EndpointRoutingDisabled {
+			endpointSlices := resources.GetEndpointSlicesForService(backendNamespace, string(backendRef.Name))
+
+			for _, endpointSlice := range endpointSlices {
+				for _, endpoint := range endpointSlice.Endpoints {
+					for _, endpointPort := range endpointSlice.Ports {
+						// Check if the endpoint port matches the service port
+						// and if endpoint is Ready
+						if *endpointPort.Name == servicePort.Name &&
+							*endpointPort.Protocol == servicePort.Protocol &&
+							*endpoint.Conditions.Ready {
+							for _, address := range endpoint.Addresses {
+								var ep *ir.DestinationEndpoint
+								// Weights are not relevant for TCP and UDP Routes
+								if routeType == KindTCPRoute ||
+									routeType == KindUDPRoute {
+									ep = ir.NewDestEndpoint(
+										address,
+										uint32(servicePort.TargetPort.IntVal))
+								} else {
+									ep = ir.NewDestEndpointWithWeight(
+										address,
+										uint32(servicePort.TargetPort.IntVal),
+										weight)
```

Comment:
I think the `weight` is not correct here.

Assume we have a route rule consisting of two backends:

```yaml
  rules:
    - backendRefs:
        - group: ""
          kind: Service
          name: foo
          port: 3000
          weight: 50
        - group: ""
          kind: Service
          name: bar
          port: 3000
          weight: 50
```

The "foo" service has 100 endpoints and the "bar" service has 500 endpoints. The actual traffic ratio between foo and bar is 2:1 instead of the intended 1:1.

Reply:
thanks for catching this, fixing this rn